### PR TITLE
Oidc integration

### PR DESCRIPTION
**Pull request by Matthew-Beliveau**
This should be the cleanest version of what I have right now. Let me know if I was missing something or there is some redundant code! Also sorry if there is any white space changes. I tried my best with the last few commits to not have any show up!

### CODE REVIEW

File: bin/keycloak-httpd-client-install [hunk 1]
```diff
@@ -1002,16 +1030,19 @@ def main():
                        help='certficate file')
 
     group.add_argument('--mellon-hostname', default=socket.getfqdn(),
-                       help="Machine's fully qualified host name")
+                       help="Deprecated. Use --hostname instead. ""
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: Syntax Error: unmatched quote due to extra double quote at end of line.

File: bin/keycloak-httpd-client-install [hunk 2]
```diff
@@ -943,6 +966,11 @@ def main():
                        default='/etc/httpd',
                        help='Template location')
 
+    group.add_argument('--module',
+                        required=True,
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: This shouldn't be required because it breaks backwards compatibility, instead it should default to mod_auth_mellon.

File: bin/keycloak-httpd-client-install [hunk 3]
```diff
@@ -1090,47 +1132,71 @@ def main():
...
-            mellon_https_port=options.mellon_https_port))
-    options.mellon_endpoint_path = join_path(options.mellon_root, options.mellon_endpoint)
-
-    if not options.mellon_entity_id:
-        url = urlsplit(options.mellon_http_url)
-        options.mellon_entity_id = urlunsplit((url.scheme, url.netloc,
-                                               join_path(options.mellon_endpoint_path, 'metadata'),
-                                               '', ''))
+    if options.module == 'mod_auth_mellon':
+        options.httpd_dir = os.path.join(options.httpd_dir, HTTPD_DIR)
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: initializing options.httpd_dir is common between modules, it should not be trapped inside a module if condition.

File: bin/keycloak-httpd-client-install [hunk 4]
```diff
@@ -1090,47 +1132,71 @@ def main():
...
+        options.httpd_conf_dir = os.path.join(options.httpd_dir, HTTPD_CONF_DIR)
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: You're initializing exactly the same way in both the mellon and oidc cases, the initialization should be common and moved out of the module condition in both blocks

File: bin/keycloak-httpd-client-install [hunk 5]
```diff
@@ -1090,47 +1132,71 @@ def main():
...
+        options.mellon_endpoint_path = join_path(options.mellon_root, options.mellon_endpoint)
+
+        if not options.mellon_entity_id:
+            url = urlsplit(options.mellon_http_url)
+            options.mellon_entity_id = urlunsplit((url.scheme, url.netloc,
+                                                   join_path(options.mellon_endpoint_path, 'metadata'),
+                                                   '', ''))
+
+    elif options.module == 'mod_auth_openidc':
+        options.httpd_conf_dir = os.path.join(options.httpd_dir, HTTPD_CONF_DIR)
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: See above comment concerning duplicate initialization.

File: bin/keycloak-httpd-client-install (hunk 2 above)
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: This shouldn't be required because it breaks backwards compatibility, instead it should default to mod_auth_mellon.

File: bin/keycloak-httpd-client-install (hunk 1 above)
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: Syntax Error: unmatched quote due to extra double quote at end of line.

File: bin/keycloak-httpd-client-install [hunk 6]
```diff
@@ -43,13 +43,17 @@ BIN_TIMEOUT = "/usr/bin/timeout"
 DEV_NULL = '/dev/null'
 BIN_TIMEOUT = '/usr/bin/timeout'
 
-HTTPD_SAML_DIR = 'saml2'
+HTTPD_DIR = 'KHCI'
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: I like having a neutral directory name although I'm not sure KHCI will be recognized by anyone not familiar with the tool, and even then I'm not sure it's obvious. Think about the admin who comes along after someone first set this up using the tool. A better name might be "keycloak" (will keycloak even be recognized by another admin if they are using RH-SSO) or possibly "federation". Let's make a note of this and ask others what they think the name should be.

File: bin/keycloak-httpd-client-install [hunk 7]
```diff
@@ -913,6 +922,20 @@ def main():
                         required=True,
                         help='name of the web app being protected by mellon')
 
+    parser.add_argument('--hostname', default=socket.getfqdn(),
+                        help="Machine's fully qualified host name")
+
+    parser.add_argument('--module-root', default='/',
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: I'm not sure the endpoint root applies to mod_auth_openidc. In fact it doesn't even look like you reference it for the oidc case. Maybe we should just leave this as mellon specific.

File: bin/keycloak-httpd-client-install [hunk 8]
```diff
@@ -1002,16 +1030,19 @@ def main():
                        help='certficate file')
 
     group.add_argument('--mellon-hostname', default=socket.getfqdn(),
-                       help="Machine's fully qualified host name")
+                       help="Deprecated. Use --hostname instead. ""
+                       " Machine's fully qualified host name")
 
     group.add_argument('--mellon-https-port', default=443, type=int,
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: This comment applies to all the new common options that were derived from deprecated mellon specific options, no need to repeat this comment for each.

This isn't actually working as a common option because you're still setting and referencing the deprecated mellon variant. First take a look at the "dest" attribute for an argparse argument: [https://docs.python.org/3/library/argparse.html#dest](url). This is the option name that appears after argument parsing is complete (e.g. the argparse namespace object). So lets say the long argument is --foo, then by default you'll reference it as options.foo. But you can specify the name of the option when you call .add_argument by specifying the dest keyword. Therefore if a deprecated argument is semantically identical (e.g. mellon-hostname and hostname) you should set the dest in the deprecated argument to match the dest of it's replacement. Then only reference the new option name because both the deprecated arg and the new arg will be using the exact same parsed option name. There are numerous places where you have to fix the code for this.

Also, adding deprecated to the help is useful but only if you read the help, most people won't or have scripts that simply use the option, they'll never know the option is deprecated. Instead at run time if someone uses a deprecated option the tool need to emit a warning message which should include suggestions on how to migrate from the old to the new.

I spent a bunch of time trying to figure out the most elegant way to do this with argparse (using decorators, callbacks, adding new keyword args, `automatically` invoking the new arg, etc.) and due to architectural limitations in argparse the simplist solution I came up with is the following, edit the deprecated arg like this: 

1. add dest keyword to match new name
2. Set the action to a class that both emits a deprecation message and stores the value in dest
3. add  a new attribute to the arg object called "deprecation_help" (used by the new action class). Unfortunately you have to add the deprecation_help attribute after the arg is created, you can't add it as a keyword.

You'll need an action class for each type action, in the example snippet I'm providing I show the the standard "store" and "append" actions. Depending on what is being deprecated you might need to add another class to handle it (e.g. choice).

```python
import copy
import six

class ActionStoreDeprecated(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        deprecation_help = getattr(self, 'deprecation_help', None)
        if deprecation_help is None:
            print('arguments %s are deprecated, '
                  'no further information available' %
                  ('|'.join(self.option_strings)),
                  file=sys.stderr)
        else:
            print('arguments %s are deprecated, %s' %
                  ('|'.join(self.option_strings), deprecation_help),
                  file=sys.stderr)
        setattr(namespace, self.dest, values)


class ActionAppendDeprecated(argparse.Action):
    def __call__(self, parser, namespace, values, option_string=None):
        deprecation_help = getattr(self, 'deprecation_help', None)
        if deprecation_help is None:
            print('arguments %s are deprecated, '
                  'no further information available' %
                  ('|'.join(self.option_strings)),
                  file=sys.stderr)
        else:
            print('arguments %s are deprecated, %s' %
                  ('|'.join(self.option_strings), deprecation_help),
                  file=sys.stderr)

        if getattr(namespace, self.dest, None) is None:
            setattr(namespace, self.dest, [])
        items = copy.copy(getattr(namespace, self.dest))
        items.append(values)
        setattr(namespace, self.dest, items)

````

So here is an example of --foo being replaced by --bar:

```python
a = parser.add_argument('--foo', '-f', dest='bar', action=ActionStoreDeprecated)
a.deprecation_help = 'use --bar|-b instead'
```

File: bin/keycloak-httpd-client-install [hunk 9]
```diff
@@ -1033,10 +1064,21 @@ def main():
...
 
     group.add_argument('-l', '--mellon-protected-locations', action='append',
-                       type=arg_type_mellon_protected_location, default=[],
-                       help='Web location to protect with Mellon. '
+                       type=arg_type_module_protected_location, default=[],
+                       help='Deprecated. Use --protected-locations instead '
+                       ' Web location to protect with Mellon. '
                             'May be specified multiple times')
 
+    # ---- Argument Group "OIDC SP" ----
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: Be consistent between "SP" and "RP" usage

File: bin/keycloak-httpd-client-install [hunk 10]
```diff
@@ -1033,10 +1064,21 @@ def main():
...
+
+    # both mellon and oidc use this but idk if they should be seperate or not
+    group = parser.add_argument_group('OIDC RP')
+
+    group.add_argument('--ssl-validate-server', default='Off')
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: I presume syntactically there are only a few valid values (e.g. On, Off). Therefore don't allow someone to set a value that can't be parsed. Using a choice argument here would be appropriate to prevent generating a syntactically invalid file.

File: bin/keycloak-httpd-client-install [hunk 11]
```diff
@@ -1090,47 +1132,71 @@ def main():
                          ('keycloak-admin-password'))
 
     # ===== Normalize Options =====
-
-    options.mellon_root = '/' + options.mellon_root.strip(' /')
-    options.mellon_endpoint = options.mellon_endpoint.strip(' /')
+    if options.module_root:
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: See above, this is probably mellon specific

File: bin/keycloak-httpd-client-install [hunk 12]
```diff
@@ -1218,39 +1289,54 @@ def main():
                                           keep_trailing_newline=True,
                                           undefined=jinja2.StrictUndefined,
                                           loader=jinja2.FileSystemLoader(
-                                              options.template_dir))
+                                              'templates'))
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: Why was this changed from options.template_dir?

File: bin/keycloak-httpd-client-install [hunk 13]
```diff
@@ -1218,39 +1289,54 @@ def main():
...
+                traceback.print_exc()
+            print('%s: %s' % (e.__class__.__name__, six.text_type(e)),
+                  file=sys.stderr)
+            return STATUS_OPERATION_ERROR
+
+    elif options.module == 'mod_auth_openidc':
+    # ===== Build OIDC metadata =====
+        try:
+            logger.step('Build OIDC SP metadata file')
+            oidc_metadata = build_mellon_sp_metadata_file(options)
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: A function called "build_mellon_sp_metadata_file" should not be producing a oidc metadata. Either fix the name or have separate functions, see above comment.

File: bin/keycloak-httpd-client-install [hunk 14]
```diff
@@ -829,24 +833,29 @@ def build_template_params(options):
 
 def build_mellon_httpd_config_file(options):
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: A function whose name includes "mellon" should not be producing data applicable to both mellon and oidc.

File: bin/keycloak-httpd-client-install [hunk 15]
```diff
@@ -829,24 +833,29 @@ def build_template_params(options):
...
     template_params = build_template_params(options)
-    template = template_env.get_template(MELLON_HTTPD_CONFIG_TEMPLATE)
+    if options.module == 'mod_auth_mellon':
+        template = template_env.get_template(MELLON_HTTPD_CONFIG_TEMPLATE)
+    elif options.module == 'mod_auth_openidc':
+        template = template_env.get_template(OIDC_MELLON_HTTPD_CONFIG_TEMPLATE)
     return template.render(template_params)
 
-
 def build_mellon_sp_metadata_file(options):
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: A function whose name includes "mellon" should not be producing data applicable to both mellon and oidc.

**Review by jdennis — COMMENTED** [review]
I've left several specific comments, but here are two things which need to be done overall:

1. You've got many commits, most of which are the result of your own development activity. Those are mostly meaningless to preserve in the code history nor are they relevant to a reviewer. Commits should be independent and logically grouped so they can be reviewed and applied independently. If it's not easy or sensible to partition the commits into individual changesets then squash everything into a single commit (this will be easier in this situation, but in the future you should think about grouping changes into smaller logical chunks.

2. There should never be whitespace only changes and new commits should never introduce any trailing whitespace. After you squash do a git diff and look for whitespace only changes and fix them.

During this process git interactive merge and git commit --amend will be useful tools at your disposal, be sure you're familiar with these.

Thank you for the good work, overall I'm pretty happy with what you've done. Thanks.

File: bin/keycloak-httpd-client-install (hunk 12 above)
Comment (Matthew-Beliveau) [reply bin/keycloak-httpd-client-install]: This change was made to make the tool work for my specifically. I forgot to change it back before commiting, sorry! I will change it back!

File: bin/keycloak-httpd-client-install [hunk 16]
```diff
@@ -1090,47 +1205,68 @@ def main():
...
+        options.mellon_endpoint_path = join_path(options.mellon_root, options.mellon_endpoint)
+
+        if not options.mellon_entity_id:
+            url = urlsplit(options.mellon_http_url)
+            options.mellon_entity_id = urlunsplit((url.scheme, url.netloc,
+                                                   join_path(options.mellon_endpoint_path, 'metadata'),
+                                                   '', ''))
+
+    elif options.module == 'mod_auth_openidc':
+        options.mellon_httpd_config_filename = \
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: This is the mod_auth_openidc section, why is the argument mellon_httpd_config_filename? Is there any reason both mellon and oidc can't use the same configuration variable, perhaps module_httpd_config_filename

File: bin/keycloak-httpd-client-install [hunk 17]
```diff
@@ -1090,47 +1205,68 @@ def main():
...
+            os.path.join(options.httpd_conf_dir,
+                         '{app_name}_oidc.conf'.format(
+                             app_name=options.app_name))
+        options.oidc_sp_metadata_filename = \
+            os.path.join(options.httpd_dir,
+                         '{app_name}_{oidc_metadata}').format(
+                             app_name=options.app_name,
+                             oidc_metadata=OIDC_METADATA)
+        options.mellon_dst_idp_metadata_file = \
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: same issue as above, you're modifying a mellon variable inside an oidc block

File: bin/keycloak-httpd-client-install [hunk 18]
```diff
@@ -1370,6 +1539,21 @@ def main():
               file=sys.stderr)
         return STATUS_OPERATION_ERROR
 
+    # ==== build OIDC Template =====
+    if options.module == 'mod_auth_openidc':
+
+        try:
+            logger.step('Build Mellon httpd config file')
```
Comment (jdennis) [inline bin/keycloak-httpd-client-install]: It's not mellon if you're in an oidc block

File: keycloak_httpd_client/keycloak_cli.py [hunk 19]
```diff
@@ -788,6 +845,11 @@ def do_register_client(options, conn):
         options.initial_access_token,
         options.realm_name, metadata)
 
+def do_register_oidc_client(options, conn):
```
Comment (jdennis) [inline keycloak_httpd_client/keycloak_cli.py:848]: I don't believe this is called anywhere, did you mean to add it to the testing options?

**Review by jdennis — COMMENTED** [review]
Overall it's looks good Mathew, but there are a few things I want you to address mostly having to do with style.

It's difficult to review a pull request that has multiple commits. Please squash all your changes into a single commit so it's easier to see the entirety of the changes. In general reviewers don't want to see your intermediate steps, they want to see the final result.

Also, there are way too many whitespace changes, please find and revert those changes. Updates to source should only include functional code changes, not differences in formatting (the later obscures the intent of the change by introducing irrelevant differences).

Be careful of large blocks of if/then/else, specifically in relation to whether the module is mellon or oidc. If you see yourself doing that consider moving those large blocks into independent functions/methods so as not to obscure the readability or intent of the code.

A lot of the if module==mellon, else if module==oidc can be handled more cleanly via other mechanisms. For instance consider how the code might look if you used class hierarchy instead. This is a good example of where subclasses can provide select independent behavior. For example, suppose you had a root class called Module, and then two subclasses derived from Module, MellonModule and OPIDCModule. Then anywhere there is a difference between how the two modules operate are handled in a method override. Does that make sense? You could also use function pointers but that's just a dirty way of implementing subclass behavior.

In general code reads cleaner and is easier to maintain if instead of a lot of if/then/else to handle the different "modes" all the "mode" logic is gathered together so you're dealing only with code for one mode, that makes it much easier to understand and more robust because you're not likely to introduce a change specific to one mode in the wrong context. Another reason to keep the changes logically grouped (instead of in if/then/else blocks) is what happens when yet a third or fourth mode is added, the if/then/else blocks will just get unwieldy.

**Comment by mimendo** [comment]
Hello,
Sorry for asking, but is there any ETA for a release with OIDC support? It would be a great help to me, and I'm sure to many others.
Thank you!

**Comment by jdennis** [comment]
Closing this PR, support for mod_auth_openidc has been added in other commits.

**Comment by jdennis** [comment]
Closing this PR because support for mod_auth_openidc has been added in other commits.
